homeassistant/components/ihc/__init__.py

```diff
@@ -0,0 +1,122 @@
+
+REQUIREMENTS = ['ihcsdk==2.1.0']
+DOMAIN = 'ihc'
+
+CONF_INFO = 'info'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_URL): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_INFO): cv.boolean
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass, config):
+    """Setyp the IHC platform."""
+    from ihcsdk.ihccontroller import IHCController
+    url = config[DOMAIN].get(CONF_URL)
+    username = config[DOMAIN].get(CONF_USERNAME)
+    password = config[DOMAIN].get(CONF_PASSWORD)
+    ihc = IHCController(url, username, password)
+    ihc.info = config[DOMAIN].get(CONF_INFO)
+
+    if not ihc.authenticate():
+        _LOGGER.error("Unable to authenticate on ihc controller.")
+        return False
+
+    hass.data[DOMAIN] = IHCPlatform(ihc)
+
+    # Service functions
+
+    def set_runtime_value_bool(call):
+        """Set a IHC runtime bool value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = bool(call.data.get('value', 0))
+        ihc.set_runtime_value_bool(ihcid, value)
+
+    def set_runtime_value_int(call):
+        """Set a IHC runtime integer value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = int(call.data.get('value', 0))
+        ihc.set_runtime_value_int(ihcid, value)
+
+    def set_runtime_value_float(call):
+        """Set a IHC runtime float value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = float(call.data.get('value', 0))
+        ihc.set_runtime_value_float(ihcid, value)
+
+    descriptions = load_yaml_config_file(
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_BOOL,
+                           set_runtime_value_bool,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_BOOL])
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_INT,
+                           set_runtime_value_int,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_INT])
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_FLOAT,
+                           set_runtime_value_float,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_FLOAT])
+    return True
+
+
+class IHCPlatform:
+    """Wraps the IHCController for caching of ihc project and autosetup."""
+
+    def __init__(self, ihccontroller):
+        """Initialize the caching of the project."""
+        self.ihc = ihccontroller
+        project = self.ihc.get_project()
+        self._project = xml.etree.ElementTree.fromstring(project)
+        self._groups = self._project.findall(r'.//group')
+
+    def get_project_xml(self):
+        """Get the cached ihc project as xml."""
+        return self._project
+
+    def get_groups_xml(self):
```

Comment:
I don't think you need this method, since it's only used in this class. Just access the attribute `_groups` directly when needed.

Reply:
Done